qtpyvcp/widgets/input_widgets/mdihistory_widget.py

```diff
@@ -119,18 +127,51 @@ def set_history(self, items_list):
             self.addItem(row_item)
 
 
-        #self.addItems(history_list)
+    def heart_beat(self):
+        """Supports heart beat on the MDI History execution queue."""
+        # check if machine is idle and ready to run another command
+        if STAT.interp_state != linuxcnc.INTERP_IDLE:
+            # RS274NGC interpreter not in a state to execute
+            print
+            return
+
+        # scan for the next command to execute from bottom up.
+        list_length = self.count()-1
+        while list_length >= 0:
+            row_item = self.item(list_length)
+            row_item_data = row_item.data(MDIHistory.MDQQ_ROLE)
+
+            if row_item_data == MDIHistory.MDIQ_RUNNING:
+                # machine is in idle state so the running command is done
+                row_item.setData(MDIHistory.MDQQ_ROLE, MDIHistory.MDIQ_DONE)
+
+            elif row_item_data == MDIHistory.MDIQ_TODO:
+                cmd = str(row_item.text()).strip()
+                row_item.setData(MDIHistory.MDQQ_ROLE, MDIHistory.MDIQ_RUNNING)
+                print "Item found to execute on: {}".format(cmd)
+                issue_mdi(cmd)
+                break
+
+            list_length -= 1
 
     def initialize(self):
         """Load up starting data and set signal connections."""
         history = STATUS.mdi_history.value
         self.set_history(history)
         self.clicked.connect(self.row_clicked)
-        #STATUS.mdi_history.notify(self.set_history)
-        self.mdi_entry_widget = self.parent().findChild(MDIEntry, self.mdi_entryline_name)
-        #print 'find child: {}'.format(self.parent().findChild(MDIEntry, 'mdiEntry'))
+
+        # get handle to windows list and seach through them
+        # for the widget referenced in mdi_entryline_name
+        for win_name, obj in qtpyvcp.WINDOWS.items():
+            if hasattr(obj, self.mdi_entryline_name):
+                self.mdi_entry_widget = getattr(obj, self.mdi_entryline_name)
+
```

Comment:
Stop scanning if we found the widget
```suggestion
                break
```